src/hotspot/share/runtime/handshake.cpp

```diff
@@ -174,39 +176,42 @@ class HandshakeSpinYield : public StackObj {
   }
 };
 
-class VM_Handshake: public VM_Operation {
-  const jlong _handshake_timeout;
- public:
-  bool evaluate_at_safepoint() const { return false; }
-
- protected:
-  HandshakeOperation* const _op;
+static void handle_timeout(HandshakeOperation* op, JavaThread* target) {
+  JavaThreadIteratorWithHandle jtiwh;
 
-  VM_Handshake(HandshakeOperation* op) :
-      _handshake_timeout(millis_to_nanos(HandshakeTimeout)), _op(op) {}
+  log_error(handshake)("Handshake timeout: %s(" INTPTR_FORMAT "), pending threads: " INT32_FORMAT,
+                       op->name(), p2i(op), op->pending_threads());
 
-  bool handshake_has_timed_out(jlong start_time);
-  static void handle_timeout();
-};
+  if (target == NULL) {
+    for ( ; JavaThread* thr = jtiwh.next(); ) {
+      if (thr->handshake_state()->operation_pending(op)) {
+        log_error(handshake)("JavaThread " INTPTR_FORMAT " has not cleared handshake op: " INTPTR_FORMAT, p2i(thr), p2i(op));
+        target = thr;
+      }
+    }
+  } else {
+    log_error(handshake)("JavaThread " INTPTR_FORMAT " has not cleared handshake op: " INTPTR_FORMAT, p2i(target), p2i(op));
+  }
 
-bool VM_Handshake::handshake_has_timed_out(jlong start_time) {
-  // Check if handshake operation has timed out
-  if (_handshake_timeout > 0) {
-    return os::javaTimeNanos() >= (start_time + _handshake_timeout);
+  if (target != NULL) {
+    if (os::signal_thread(target, SIGILL, "cannot be handshaked")) {
```

Comment:
Note this is a no-op on Windows. Not sure if you want to conditionalize this or not.

Reply:
This is the same way we do it in safepoint.cpp. Also it's a bit cleaner without, so I'll keep it this way.